src/main/java/de/bonndan/nivio/output/icons/LocalIcons.java

```diff
@@ -0,0 +1,115 @@
+package de.bonndan.nivio.output.icons;
+
+import de.bonndan.nivio.model.Item;
+import de.bonndan.nivio.model.Label;
+import de.bonndan.nivio.util.URLHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
+import java.net.URL;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static de.bonndan.nivio.output.icons.IconMapping.DEFAULT_ICON;
+
+/**
+ * Provides icons and vendor icons as data.
+ *
+ *
+ */
+@Service
+public class LocalIcons {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LocalIcons.class);
+
+    private final VendorIcons vendorIcons;
+
+    /**
+     * default icon data url
+     */
+    private final String defaultIcon;
+
+    /**
+     * data url cache
+     */
+    private final Map<String, String> iconDataUrls = new ConcurrentHashMap<>();
+
+
+    public LocalIcons(VendorIcons vendorIcons) {
+
+        this.vendorIcons = vendorIcons;
+        defaultIcon = getIconUrl(DEFAULT_ICON.getIcon(), false);
+    }
+
+    public String getIconUrl(Item item) {
+
+        //icon label based
+        String icon = item.getIcon();
+        if (!StringUtils.isEmpty(icon)) {
+
+            if (icon.startsWith(VendorIcons.VENDOR_PREFIX)) {
+                String key = icon.replace(VendorIcons.VENDOR_PREFIX, "").toLowerCase();
+                return vendorIcons.getUrl(key).orElse(defaultIcon);
+            }
+
+            String iconUrl = getIconUrl(icon, true);
+            if(iconUrl != null) {
+                return iconUrl;
+            }
+        }
+
+        //type based
+        String type = item.getLabel(Label.type);
+        if (StringUtils.isEmpty(type)) {
+            return getIconUrl(DEFAULT_ICON.getIcon(), false);
+        }
+
+        //fallback to item.type
+        String iconName = IconMapping.of(type.toLowerCase()).map(IconMapping::getIcon).orElseGet(type::toLowerCase);
+        return getIconUrl(iconName, true);
+    }
+
+    /**
+     * Provides an URL for a locally served icon.
+     *
+     * @return an url pointing to a file or a data url
+     */
+    String getIconUrl(String icon, boolean fallback) {
```

Comment:
Can't this be private?

Reply:
I've been lazy and made it package-private for testing. 🙈

Reply:
But I did not see it being used in the tests - did I overlook this?

Reply:
LocalIconsTest uses it. Maybe it were better to return an Optional and get rid of the boolean flag.

Reply:
Oh, yes.

Yes, that would be better!

Reply:
done